src/extent.c

```diff
@@ -151,17 +161,27 @@ ecache_evict(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
 	edata_t *edata;
 	while (true) {
 		/* Get the LRU extent, if any. */
-		edata = edata_list_inactive_first(&ecache->eset.lru);
+		eset_t *eset = &ecache->eset;
+		edata = edata_list_inactive_first(&eset->lru);
 		if (edata == NULL) {
-			goto label_return;
+			/*
+			 * Next check if there are guarded extents.  They are
```

Comment:
Hmm, have you tried this in long-running prod canaries? My expectation would be that this would cause a slow "leaking" from the unguarded to the guarded eset over time, so that the fraction of time doing the more expensive purge operations would increase over time.

Reply:
I haven't tried the long running ones. But I have concerns over the VM fragmentation part (that I didn't mention before) -- 
1) I have another diff to add a dedicated retained space for the guarded extents, i.e. to have grow_retained for guarded as well. That way we make sure the guarded extents are from their own `chunks`, instead of mixed with regular ones.
2) The other part is, I think we should stick to the `cache guarded extents for slab sizes only` rule, so that we only have a few page sizes (4 by default, 1,3,5 or 7 pages) ever containing cached guarded extents. This should limit the amount of guarded extents around.

Like you mentioned, if we always prefer purging the regular ones first, there's a risk of the guarded ones accumulating. I'll make sure to evaluate that, probably after adding the separate grow_retained space though, since I'm fairly certainly about VM fragmentation issue.

Reply:
I'm not sure I follow -- why does this make a difference? It seems like this is a purging-policy-only issue, right?

In any case, I think this is fine "for now", it's really just a question for when we want to turn this on somewhere for real. I'm OK with it as we're still experimenting.

Reply:
Sorry I should have made it clearer -- right now I'm more worried about the other part (the VM fragmentation) that causes slowdowns over time, and it also has a side effect of making purging more expensive, because of the small punched holes by the guards which prevents effective merging as well. I wanted to resolve that part first, and then comes back to the purging policy here.

For a potential fix / better policy, we can try purging them using the "sample ratio", e.g. if the runtime option is one guarded every N extents, then we also purge 1 guarded after purging every N-1 regular ones. But yeah I prefer leaving that to a future PR.